### PR TITLE
Main server basis opgezet

### PR DESCRIPTION
**Pull request by Laugslander**
Main server applicatie in grote lijnen opgezet.

- Repository Design Pattern toegepast voor manipulatie MySQL database. 
- Database queries zoveel mogelijk geimplementeerd als stored procedures.

### CODE REVIEW

File: Soccar [Main-Server]/src/nl/soccar/library/SessionData.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    private static final int DEFAULT_CAPACITY = 6;
+
+    private final String roomName;
+    private final String hostName;
+    private final boolean hasPassword;
+    private int occupation;
+    private int capacity;
+
+    /**
+     * Constructor user for instantiation of a new SessionData object. A
```
Comment (Ranray) [inline Soccar [Main-Server]/src/nl/soccar/library/SessionData.java]: Constructor **used**

File: Soccar [Main-Server]/src/nl/soccar/library/Statistics.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+    /**
+     * Calculates the ratio of won/lost games compared to the total amount of
+     * games played. The games ratio is calculated by substracting the lost
+     * games from the won games, devided by the total number of played games and
+     * multiplied by 100.
+     *
+     * @return Calculated games ratio, higher/equal than/to 0.
+     */
+    public String getGamesRatio() {
```
Comment (Ranray) [inline Soccar [Main-Server]/src/nl/soccar/library/Statistics.java]: Waarom returned deze methode een String en niet gewoon een Double?

File: Soccar [Main-Server]/src/nl/soccar/mainserver/data/context/UserMySqlContext.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+        try {
+            CallableStatement cs = DatabaseUtilities.prepareCall("{call add_user(?, ?, ?)}");
+            cs.setString(1, username);
+            cs.setBytes(2, hash);
+            cs.setBytes(3, salt);
+            cs.execute();
+            return true;
+        } catch (SQLException e) {
+            return false;
```
Comment (Ranray) [inline Soccar [Main-Server]/src/nl/soccar/mainserver/data/context/UserMySqlContext.java:40]: Graag de exception loggen, wordt op dit moment niets mee gedaan.

File: Soccar [Main-Server]/src/nl/soccar/mainserver/data/context/UserMySqlContext.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+    @Override
+    public boolean checkIfExists(String username) {
+        try {
+            CallableStatement cs = DatabaseUtilities.prepareCall("{? = call check_user_exists(?)}");
+            cs.registerOutParameter(1, Types.BOOLEAN);
+            cs.setString(2, username);
+            cs.execute();
+            return cs.getBoolean(1);
+        } catch (SQLException e) {
```
Comment (Ranray) [inline Soccar [Main-Server]/src/nl/soccar/mainserver/data/context/UserMySqlContext.java:64]: Graag de exception loggen, wordt op dit moment niets mee gedaan.

File: Soccar [Main-Server]/src/nl/soccar/mainserver/data/repository/UserRepository.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+     *
+     * @param context The persistance service (context) that this repository
+     * communicates with.
+     */
+    public UserRepository(IUserDataContract context) {
+        this.context = context;
+    }
+
+    /**
+     * Method that adds a new user to the persistency service.
```
Comment (Ranray) [inline Soccar [Main-Server]/src/nl/soccar/mainserver/data/repository/UserRepository.java]: Adds a new user to the persistency service.

File: Soccar [Main-Server]/src/nl/soccar/mainserver/data/repository/StatisticsRepository.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+     */
+    public void incrementGamesPlayed(String username) {
+        super.getPool().execute(() -> {
+            context.incrementGamesPlayed(username);
+        });
+    }
+
+    /**
+     * Method that gets the game statistics of a player stored in the
+     * persistency service.
```
Comment (Ranray) [inline Soccar [Main-Server]/src/nl/soccar/mainserver/data/repository/StatisticsRepository.java]: Gets the game statistics of a player stored in the persistency service.

File: Soccar [Main-Server]/src/nl/soccar/mainserver/data/repository/StatisticsRepository.java [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+     */
+    public void incrementGamesWon(String username) {
+        super.getPool().execute(() -> {
+            context.incrementGamesWon(username);
+        });
+    }
+
+    /**
+     * Method that increments the amount of games lost by a player stored in the
+     * persistency service.
```
Comment (Ranray) [inline Soccar [Main-Server]/src/nl/soccar/mainserver/data/repository/StatisticsRepository.java]: Increments the amount of games lost by a player stored in the persistency service.

File: Soccar [Main-Server]/src/nl/soccar/mainserver/data/repository/StatisticsRepository.java [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+     */
+    public void addAssists(String username, int assists) {
+        super.getPool().execute(() -> {
+            context.addAssists(username, assists);
+        });
+    }
+
+    /**
+     * Method that increments the amount of games won by a player stored in the
+     * persistency service.
```
Comment (Ranray) [inline Soccar [Main-Server]/src/nl/soccar/mainserver/data/repository/StatisticsRepository.java]: Increments the amount of games won by a player stored in the persistency service.

File: Soccar [Main-Server]/src/nl/soccar/mainserver/data/repository/StatisticsRepository.java [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+     */
+    public void addGoals(String username, int goals) {
+        super.getPool().execute(() -> {
+            context.addGoals(username, goals);
+        });
+    }
+
+    /**
+     * Method that adds a given amount of assists to the given player stored in
+     * the persistency service.
```
Comment (Ranray) [inline Soccar [Main-Server]/src/nl/soccar/mainserver/data/repository/StatisticsRepository.java]: Adds a given amount of assists to the given player stored in the persistency service.

File: Soccar [Main-Server]/src/nl/soccar/mainserver/data/repository/StatisticsRepository.java [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+     * @param context The persistance service (context) that this repository
+     * communicates with.
+     */
+    public StatisticsRepository(IStatisticsDataContract context) {
+        this.context = context;
+    }
+
+    /**
+     * Method that adds a given amount of goals to the given player stored in
+     * the persistency service.
```
Comment (Ranray) [inline Soccar [Main-Server]/src/nl/soccar/mainserver/data/repository/StatisticsRepository.java]: Adds a given amount of goals to the given player stored in the persistency service.

File: Soccar [Main-Server]/src/nl/soccar/mainserver/data/repository/StatisticsRepository.java [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+     */
+    public void incrementGamesLost(String username) {
+        super.getPool().execute(() -> {
+            context.incrementGamesLost(username);
+        });
+    }
+
+    /**
+     * Method that increments the amount of games played by a player stored in
+     * the persistency service.
```
Comment (Ranray) [inline Soccar [Main-Server]/src/nl/soccar/mainserver/data/repository/StatisticsRepository.java]: Increments the amount of games played by a player stored in the persistency service.

File: Soccar [Main-Server]/src/nl/soccar/mainserver/data/repository/StatisticsRepository.java [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+            return f.get();
+        } catch (InterruptedException | ExecutionException e) {
+            LOGGER.warn("An error occurred while submitting a callable in the getStatistics method.", e);
+            return null;
+        }
+    }
+
+    /**
+     * Method that gets the game statistics of all players from the persistency
+     * service.
```
Comment (Ranray) [inline Soccar [Main-Server]/src/nl/soccar/mainserver/data/repository/StatisticsRepository.java]: Gets the game statistics of all players from the persistency service.

File: Soccar [Main-Server]/src/nl/soccar/mainserver/data/repository/UserRepository.java [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+            return f.get();
+        } catch (InterruptedException | ExecutionException e) {
+            LOGGER.warn("An error occurred while submitting a callable in the add method.", e);
+            return false;
+        }
+    }
+
+    /**
+     * Method that changes the privilege of the given player stored in the
+     * persistency service.
```
Comment (Ranray) [inline Soccar [Main-Server]/src/nl/soccar/mainserver/data/repository/UserRepository.java]: Changes the privilege of the given player stored in the persistency service.

File: Soccar [Main-Server]/src/nl/soccar/mainserver/data/repository/UserRepository.java [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+     */
+    public void changePrivilege(String username, Privilege privilege) {
+        super.getPool().execute(() -> {
+            context.changePrivilege(username, privilege);
+        });
+    }
+
+    /**
+     * Method that checks if a player exists in the persistency service based on
+     * the stored username.
```
Comment (Ranray) [inline Soccar [Main-Server]/src/nl/soccar/mainserver/data/repository/UserRepository.java]: Checks if a player exists in the persistency service based on the stored username.

File: Soccar [Main-Server]/src/nl/soccar/mainserver/data/repository/UserRepository.java [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+            return f.get();
+        } catch (InterruptedException | ExecutionException e) {
+            LOGGER.warn("An error occurred while submitting a callable in the checkIfExists method.", e);
+            return false;
+        }
+    }
+
+    /**
+     * Method that checks if the given hashed password belongs to the given
+     * username stored in the persistency service.
```
Comment (Ranray) [inline Soccar [Main-Server]/src/nl/soccar/mainserver/data/repository/UserRepository.java]: Checks if the given hashed password belongs to the given username stored in the persistency service.

File: Soccar [Main-Server]/src/nl/soccar/mainserver/rmi/server/MainServer.java [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+     * Gets the user data repository.
+     *
+     * @return The user data repository.
+     */
+    public UserRepository getUserRepository() {
+        return userRepository;
+    }
+
+    /**
+     * gets the statistics data repository.
```
Comment (Ranray) [inline Soccar [Main-Server]/src/nl/soccar/mainserver/rmi/server/MainServer.java]: **G**ets the statistics data repository.

File: Soccar [Main-Server]/src/nl/soccar/mainserver/util/DatabaseUtilities.java [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+                LOGGER.info("Database connection closed.");
+            } catch (SQLException e) {
+                LOGGER.error("An error occurred while closing the database connection.", e);
+            }
+        }
+    }
+
+    /**
+     * Method that prepares an SQL query call to the database using the active
+     * database connection.
```
Comment (Ranray) [inline Soccar [Main-Server]/src/nl/soccar/mainserver/util/DatabaseUtilities.java]: Prepares an SQL query call to the database using the active database connection.

File: Soccar [Main-Server]/src/nl/soccar/mainserver/util/DatabaseUtilities.java [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+     * @throws SQLException The exception that can be thrown when the SQL query
+     * fails.
+     */
+    public static CallableStatement prepareCall(String query) throws SQLException {
+        return connection.prepareCall(query);
+    }
+
+    /**
+     * Method that prepares an SQL query statement to the database using the
+     * active database connection.
```
Comment (Ranray) [inline Soccar [Main-Server]/src/nl/soccar/mainserver/util/DatabaseUtilities.java]: Prepares an SQL query statement to the database using the active database connection.

File: Soccar [Main-Server]/src/nl/soccar/mainserver/util/PasswordUtilities.java [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+    /**
+     * Constructor that is intentionally marked private so a PasswordUtilities
+     * object can never be initiated outside this class.
+     */
+    private PasswordUtilities() {
+    }
+
+    /**
+     * Methods that generates a random salt of 16 bytes based on the SHA1PRNG
+     * algoritm.
```
Comment (Ranray) [inline Soccar [Main-Server]/src/nl/soccar/mainserver/util/PasswordUtilities.java]: Generates a random salt of 16 bytes based on the SHA1PRNG algoritm.

File: Soccar [Main-Server]/src/nl/soccar/mainserver/util/PasswordUtilities.java [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+        } catch (NoSuchAlgorithmException e) {
+            LOGGER.error("An error occurred while generating a salt.", e);
+        }
+
+        return salt;
+    }
+
+    /**
+     * Method that adds a salt to a hashed password and combines them as one
+     * byte array of 64 bytes.
```
Comment (Ranray) [inline Soccar [Main-Server]/src/nl/soccar/mainserver/util/PasswordUtilities.java]: Adds a salt to a hashed password and combines them as one byte array of 64 bytes.

**Review by Ranray — CHANGES_REQUESTED** [review]

**Comment by Laugslander** [comment]
Gevraagde veranderingen doorgevoerd.

**Review by Ranray — APPROVED** [review]

**Pull request merged**